File: Sorting Plateau/Counting Sort_Merge Sort_Quickselect/Q3. Insertion Sort List/Q3_InsertionSortList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def insertionSortList(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        # Create a dummy node to simplify insertions at head
        dummy = ListNode(0)
        
        # Pointer to traverse the original list
        curr = head
        
        # Traverse all nodes in the original list
        while curr:
            # Start from dummy to find insertion point
            prev = dummy
            
            # Find the correct position to insert current node
            while prev.next and prev.next.val < curr.val:
                prev = prev.next
            
            # Store next node before changing links
            next_temp = curr.next
            
            # Insert current node in sorted position
            curr.next = prev.next
            prev.next = curr
            
            # Move to next node in original list
            curr = next_temp
        
        # Return the sorted list starting after dummy
        return dummy.next

File: Sorting Plateau/Counting Sort_Merge Sort_Quickselect/Q3. Insertion Sort List/Q3_InsertionSortList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        // Dummy node to simplify insertions at head
        ListNode dummy = new ListNode(0);
        
        // Start from the head of original list
        ListNode curr = head;
        
        // Traverse all nodes in the original list
        while (curr != null) {
            // Start from dummy for finding insertion point
            ListNode prev = dummy;
            
            // Find where to insert current node in sorted list
            while (prev.next != null && prev.next.val < curr.val) {
                prev = prev.next;
            }
            
            // Store next node before changing links
            ListNode nextTemp = curr.next;
            
            // Insert current node in sorted position
            curr.next = prev.next;
            prev.next = curr;
            
            // Move to next node in original list
            curr = nextTemp;
        }
        
        // Return sorted list starting after dummy
        return dummy.next;
    }
}
